README.md
### LindenmayeR
#### Functions to Explore L-Systems (Lindenmayer Systems)

`LindenmayeR is a WORK IN PROGRESS`.  The fundamental structures & behavior may change at any time.  Check the NEWS file.

`LindenmayeR` currently implements 2-dimensional, deterministic, context-free L-systems, or DOLs.  Wikipedia has a very good introduction [page](https://en.wikipedia.org/wiki/L-system) to the topic.

Numerous implementations of L-systems can be found on the web (github has plenty).  I believe this is the first one in `R` and it is intended as an exploratory and learning tool.

At this time, I am trying to make sure that the basic actions exist and that everything works as desired by implementing some examples from the Wikipedia page, as well as the material in the early pages of the book by [Prusinkiewicz and Lindenmayer](http://algorithmicbotany.org/papers/#abop).  Insofar as possible I am trying to use standard conventions.  Current examples include simple fractals and some basic plant-like structures.

The long term goal is to implement awesome 3D examples.  I'm in no particular hurry.  This interests me because the diagrams simulating plant growth are beautiful and interesting.  Long after Lindenmayer's work, it appears that the growth forms of plants work on a molecular level exactly as he imagined, via homebox type genes.  See Lippman ZB, Cohen O, Alvarez JP, Abu-Abied M, Pekker I, *et al.* (2008) The making of a compound inflorescence in tomato and related nightshades. [PLoS Biol 6(11): e288](http://dx.doi.org/10.1371/journal.pbio.0060288)

Here is the coordinate system used in the 3D implementation.

![3D Coordinate System](CoordSystem.png)

#### License Information

`LindenmayeR` is distributed under the GPL-3 license, as stated in the DESCRIPTION file.  For more info, see the [GPL site.](https://gnu.org/licenses/gpl.html)

Questions?  hanson@depauw.edu

#### Lagniappe

![](FractalSeaweed.png)


NEWS

NEWS file for package LindenmayeR

Lindenmayer or L-systems are parallel rewriting systems which can be used to simulate biological forms and certain kinds of fractals.  Briefly, in an L-system a series of symbols in a string are replaced iteratively according to rules to give a more complex string composed of the same symbols. Eventually, the symbols are translated into turtle graphics for plotting.  Wikipedia has a very good introduction: en.wikipedia.org/wiki/L-system This package provides basic functions for exploring L-systems.

Bryan A. Hanson DePauw University, Greencastle Indiana USA
URL: github.com/bryanhanson/LindenmayeR

Changes in version 0.1.14 2019-06-05
     + Checked against R 3.6.
     + In Lsys verbose = 0L is now completely quiet (fix by Martin Pedersen).

Changes in version 0.1.13 2017-07-31
     + Checked against R 3.5 RC.
     + Improved documentation.

Changes in version 0.1.6 2015-07-16
     + Updated for R 3.3 RC. No changes were needed, so really just a check that it runs in the new environment.

Changes in version 0.1.4 2015-04-16
     + Update for new version of stringr & fixed a few fiddly details.

Changes in version 0.1-3 2014-12-12
     + Rolled back the changes for stringr; there was a bug in the development version of stringr.

Changes in version 0.1-2 2014-12-10
     + Updated to work with dev version of stringr, thanks to Hadley for the heads up.
     + 3D functions removed from this version temporarily while I sort out CRAN vs Github.

Changes in version 0.1-1 2014-12-10
     + Added image to README.md
     + Some work done towards 3D drawing, but it's not right.  Don't trust!
     + Has not been built or checked.
     + Not pushed to public repo.

Changes in version 0.1-0 2014-08-11
     + First version with basic functionality.


R/Lsys.R
##' Rewrite an Axiom Using Production Rules to Give a String Ready for Turtle Graphics
##'
##' This is the central function for rewriting an initial string of symbols (the axiom)
##' into a new string using production rules.  Production rules are very simple: if
##' the symbol is A, turn it into something.  If it is B, turn it into something else.
##' Production rules typically contain instructions about moving while drawing,
##' moving w/o drawing, changing direction, or storing the current state for re-use later.
##'
##' The job of this function is to take an input "axiom" and apply the "production
##' rules" and other paramters to create a new string of drawing instructions.  The
##' "language" or character set of the axiom and production rules are arbitary, and
##' the internet and literature contains many different examples.  The same fractal
##' could be drawn using completely different sets of symbols.  The string produced
##' by this function is processed by \code{\link{drawLsys}}.  See there for further
##' explanation and examples.
##'
##' @param init A character string giving variables (symbols) to use as the initial string
##' Also known as the axiom.
##'
##' @param rules A data frame containing columns "inp" and "out".  These contain the input
##' variables and the corresponding replacement string.  See the examples in
##' \code{\link{drawLsys}}.
##'
##' @param n An integer giving the number of cycles or iterations desired.
##'
##' @param retAll Logical.  If \code{TRUE}, the result at each cycle will be returned,
##' otherwise only the last result is returned.
##'
##' @param verbose An integer giving the level of information desired as the calculation
##' proceeds.  \code{verbose = 1L} gives basic information at each cycle.
##' Any value greater than 1 gives much more detail.  Supress messages by using a value
##' less than 1.
##'
##' @return If \code{retAll = FALSE}, a character vector of length 1 giving the string
##' at the end processing.  Otherwise, a character vector of length \code{n + 1}
##' containing \code{init} plus the results at the end of each iteration.
##' 
##' @name Lsys
##' @rdname Lsys
##' @export
##' @keywords utilities
##'
##' @seealso \code{\link{drawLsys}} for examples, including plotting.
##'

Lsys <- function(init = NULL, rules = NULL, n = 5,
	retAll = TRUE, verbose = 1L) {

	nc <- nchar(rules$inp)
	if (any(nc > 1)) stop("Input variables must be a single character")
	
	if (verbose == 1L) cat("\nCycle 0 string has length ", nchar(init), "\n", sep = "")
	if (verbose == 1L) cat("Cycle 0:", init, "\n")
	curr <- init
	out <- rep(NA_character_, n+1) # save init and all output
	out[1] <- init
	
	for (j in 1:n) {
		# apply all the rules simultaneously to the current string
		# this is different than any built-in string processing
		# I can find in R
		
		RR <- vector("list") # save the rule output here
		for (i in 1:nrow(rules)) {
			rr <- str_locate_all(curr, rules[i,1])
			if (verbose > 1L) cat("Processing rule", i, "\n")
			if (dim(rr[[1]])[1] == 0) {
				if (verbose > 1L) cat("\tRule", i, "was not needed\n")
				next
				}		
			RR[i] <- rr
			}
			
		# reformat RR into something more useful
		if (verbose >= 1L) print(RR)
		
		
		for (i in 1:length(RR)) {
			tmp <- as.data.frame(RR[i])
			# if (is.na(tmp[1,1])) {
				
				# next # empty output as rule didn't apply
				# }
			tmp$insert <- rules[i,2]
			RR[[i]] <- tmp
			}

		if (verbose >= 1L) print(RR)
		
		# RR is still a list, unlist to a data frame
		
		RRdf <- as.data.frame(RR[1])
		for (i in 2:length(RR)) {
			RRdf <- rbind(RRdf, as.data.frame(RR[i]))
			}
			
		if (verbose > 1L) print(RRdf)
		
		# assemble a new string using the rules
		
		curr <- unlist(strsplit(curr, ""))
		curr[RRdf$start] <- RRdf$insert
		curr <- paste0(curr, collapse = "")
		out[j+1] <- curr
		if (verbose == 1L) cat("\nCycle ", j, " string has length ", nchar(curr), "\n", sep = "")
		if (verbose == 1L) cat("Cycle ", j, ": ", curr, "\n", sep = "")
		}
		
	if (retAll) return(out)
	curr
	}
	


R/ggLsys.R
##' S3 method to draw a 2D L-System with ggplot2
##'
##' This function takes either returned dataframe from Lsys from \code{\link{recLsys}}
##' or initial strings to create which, to create plots with ggplot
##'
##' @param x A character vector, or dataframe containing all required info to plot Lsys. For
##' details, see \code{\link{drawLsys}}
##'
##'
##' @return a gg object, by default it would be plotted.
##'
##' @name ggLsys
##' @rdname ggLsys
##' @export
##' @keywords plot
##'



ggLsys=function(x,...) UseMethod("ggLsys")


ggLsys.data.frame = function(x,...){
  requireNamespace('ggplot2')
  alpha= rep(1,(nrow(x)))#color all paths with no transparency
  index_rev = which(as.vector.factor(x$pType)=="r")

  index_beforeRev = index_rev-1
  for(i in index_beforeRev) alpha[i] = .1

  x$alpha = alpha

  ggplot(data =x,aes(x = x,y = y,
                           alpha = alpha)) + geom_path()
}

ggLsys.character = function(x, drules = NULL,
                                 st = c(5, 50, 0), stepSize = 1.0, ang = 90.0,
                                 which = length(string), shrinkFactor = NULL){
  #x = recLsys(string=x,drules,st,stepSize,ang,which,shrinkFactor)

  x =do.call(recLsys,list(string = x,drules=drules,st=st,stepSize=stepSize,ang=ang,
                          which = which,shrinkFactor=shrinkFactor))
  ggLsys.data.frame(x)

}


R/move3D.R
##' Translate a point in 3D
##'
##' This function translates a point in 3D, given a starting point, a current heading,
##' and a step size. Called by \code{\link{drawLsys3D}} and not normally by the user.
##'
##' @param cp A numeric vector of length 3 giving the current x, y, z coordinates of the point.
##'
##' @param ch A numeric vector of length 3 giving the current heading, described by
##' three angles H, L, U, given in degrees.  See Details.
##'
##' @param stepSize Numeric.  The length of the drawing step.
##'
##' @return A numeric vector of length 3 giving the new x, y, z coordinates for the point.
##'
##' @section Details The assumption of an L-system is that only one thing is done at a
##' time.  Thus, this function give correct answers only if two of H, L and U are zero.
##' If two values were non-zero, the answer would depend upon the order of operations.
##' This situation is even more restrictive however, since the value of H is ignored - it
##' corresponds to a barrel roll and does not affect the translation to the new position.
##' See the README for the coordinate system.
##'
##' @name move3D
##' @rdname move3D
##' @export
##' @keywords utilities
##'

move3D <- function(cp, ch, stepSize) {

	# Check that either L or U is zero (H is ignored)
	LU <- floor(abs(ch[2:3]))
	print(LU)
	if ((LU[1] > 0L) & (LU[2] > 0L)) stop("Either L or U must be zero")
	
	# Using spherical coordinates to Cartesian coordinate eqns
	# based on the idea of a unit sphere
	
	# H <- ch[1]*pi/180 # not used: corresponds to a barrel roll
	L <- ch[2]*pi/180
	U <- ch[3]*pi/180
	x <- stepSize*cos(U)*cos(L)
	y <- stepSize*sin(U)*cos(L)
	z <- stepSize*sin(L)

	#print(x^2 + y^2 + z^2) # check
	
	return(c(x, y, z) + cp) # returns updated cp
	}


# some tests
#             x, y, z          H, L, U
# move3D(cp = c(0, 0, 0), ch = c(0, 0, 0), stepSize = 1.0)
# move3D(cp = c(0, 0, 0), ch = c(90, 0, 0), stepSize = 1.0)
# move3D(cp = c(0, 0, 0), ch = c(0, 0, 90), stepSize = 1.0)
# move3D(cp = c(0, 0, 0), ch = c(0, 0, -90), stepSize = 1.0)
# move3D(cp = c(0, 0, 0), ch = c(0, 90, 0), stepSize = 1.0)
# move3D(cp = c(0, 0, 0), ch = c(0, -90, 0), stepSize = 1.0)

# move3D(cp = c(0, 0, 0), ch = c(0, 45, 0), stepSize = 1.0)
# move3D(cp = c(0, 0, 0), ch = c(0, 0, 45), stepSize = 1.0)
# move3D(cp = c(0, 0, 0), ch = c(45, 0, 0), stepSize = 1.0)

# move3D(cp = c(0, 0, 0), ch = c(45, 90, 90), stepSize = 1.0) # error trapping



R/recLsys.R
##' Record path of the virtual pen with turtle graphics from Lsys
##'
##' This function takes input strings, previously created with \code{\link{Lsys}},
##' translates them to coordinates and types of line segments stored in a data frame.
##'
##' @param string A character vector, or dataframe containing all required info to plot Lsys. For
##' details, see \code{\link{drawLsys}}
##'
##'
##' @return a data frame, containing coordinate data and segment types
##'   s for start of a segment d for segment endpoints to be drawn, r for reverse points#'
##'
##' @name recLsys
##' @rdname recLsys
##' @export
##' @keywords coordinate
##'

recLsys = function(string, drules = NULL,
                            st = c(5, 50, 0), stepSize = 1.0, ang = 90.0,
                            which = length(string), shrinkFactor = NULL){

  # check drules to make sure only allowed characters were given
  OK <- c("F", "f", "+", "-", "[", "]")
  test <- drules$action
  if (!all(test %in% OK)) {
    msg1 <- paste("Only the following actions are recognized:",
                  paste(OK, collapse = " "), sep = " ")
    message(msg1)
    good <- test %in% OK
    bad <- test[!good]
    msg2 <- paste("I can't use these:",
                  paste(bad, collapse = " "), sep = " ")
    stop(msg2)
  }
  for (n in 1:length(which)) {
  string <- unlist(strsplit(string[which[n]], ""))

  for (i in 1:nrow(drules)) { # translate to drawing commands
    for (j in 1:length(string)) {
      if (string[j] == drules$symbol[i]) string[j] <- drules$action[i]
    }
  }

  # execute the drawing instructions

  #grid.move.to(st[1], st[2], default.units = "native")
  cp <- st # cp = current point
  ch <- st[3] # ch = current heading, 0 = East in degrees
  if (!is.null(shrinkFactor)) stepSize <- stepSize/shrinkFactor
  fifo <- vector("list") # store info to restore later
  ns <- 0L # stack counter


  xRec = cp[1,drop =T]
  yRec = cp[2,drop = T]

  typeRec = 's' # type of coord data
  # s - start
  # d - draw
  # r - reverse
  for (j in 1:length(string))	{
    #cat("Processing character", j, "\n")
    if (string[j] == "F") {
      x <- cp[1] + stepSize*cos(ch*pi/180)
      y <- cp[2] + stepSize*sin(ch*pi/180)
      #
      cp <- c(x,y)
      xRec = c(xRec,x)
      yRec = c(yRec,y)
      typeRec <- c(typeRec,"d")
    }else	if(string[j] == "f") {
      x <- cp[1] + stepSize*cos(ch*pi/180)
      y <- cp[2] + stepSize*sin(ch*pi/180)
      cp <- c(x,y)

      xRec = c(xRec,x)
      yRec = c(yRec,y)
      typeRec<-c(typeRec,"m")
    }else if(string[j] == "[") {
      #cat("Found a [ \n")
      #cat("ns is:", ns, "\n")
      ns <- ns + 1 # save the current settings
      fifo[[ns]] <- c(cp, ch)
      #print(fifo)
    }else if(string[j] == "]") {
      #cat("Found a ] \n")
      #cat("ns is:", ns, "\n")
      cp <- fifo[[ns]][1:2]
      ch <- fifo[[ns]][3]

      ns <- ns - 1

      xRec = c(xRec,cp[1])
      yRec = c(yRec,cp[2])
      typeRec = c(typeRec,'r')
      #print(fifo)
    }else	if(string[j] == "-"){
      ch = ch - ang
    }else{
      if (string[j] == "+") ch = ch + ang
    }
  }
  }
  return(data.frame(x = xRec,y = yRec,pType = typeRec))
}


R/drawLsys3D.R
##' Draw a 3D L-System Using Turtle Graphics
##'
##' This function takes input strings, previously created with \code{\link{Lsys}},
##' translates them into 3D turtle graphics instructions, and then plots the results.
##' 
##' @param string A character vector giving the strings containing the turtle graphics
##' instructions.  Created by \code{\link{Lsys}}.
##'
##' @param drules A data frame containing columns "symbols" and "action".  These contain the input
##' symbols and the corresponding drawing action.  See the examples.
##'
##' @param st A numeric vector of length 3 giving the screen coordinates where
##' the start of the curve should be placed.  The screen is 100 x 100 with the
##' lower left corner as 0,0.  The third element is the initial drawing angle
##' in degrees.
##'
##' @param stepSize Numeric.  The length of the drawing step.
##'
##' @param ang Numeric.  The angle in degrees when a change in direction is requested.
##'
##' @param which Integer.  The entries in \code{string} which should be drawn.  Defaults
##' to the last (most complex) entry.  If \code{length(which) > 1} each plot is drawn in
##' its own window.
##'
##' @param shrinkFactor A numeric vector of the same length as \code{string}.  As each
##' plot is made, \code{stepSize} will be divided by the corresponding value in \code{shrinkFactor}.
##' This allows one to scale down the increasingly large/complex plots to make them
##' occupy a space similar to the less complex plots.
##'
##' @param ...  Additional parameters to be passed to the \code{grid} drawing routines.
##' Most likely, something of the form \code{gp = gpar(...)}.  See \code{\link{gpar}}
##' and the last example.
##'
##' @section Warning: Remember that if \code{retAll = TRUE}, \code{\link{Lsys}} returns the initial string plus the results
##' of all iterations.  In this case, if you want the 5th iteration, you should specify \code{which = 6} since
##' the initial string is in \code{string[1]}.
##' 
##' @return None; side effect is a plot.
##' 
##' @name drawLsys3D
##' @rdname drawLsys3D
##' @export
##' @keywords plot
##'


drawLsys3D <- function(string = NULL, drules = NULL,
	stepSize = 1.0, angle = 45,
	which = length(string), ...) {

	# check drules to make sure only allowed characters were given
	OK <- c(LETTERS, letters, "+", "-", "[", "]", "^", "&", "\\", "/", "|")
	if (!all(test %in% OK)) {
		msg1 <- paste("Only the following actions are recognized:",
			paste(OK, collapse = " "), sep = " ")
		message(msg1)
		good <- test %in% OK
		bad <- test[!good]
		msg2 <- paste("I can't use these:",
			paste(bad, collapse = " "), sep = " ")
		stop(msg2)
		}

	for (n in 1:length(which)) {

		# convert the initial string into drawing instructions
		# F = move forward; f = move w/o drawing
		# [ save cp, ch; ] restore saved cp, ch
		# ^ = pitch up, & = pitch down (Rl)
		# \ = roll left, / = roll right (Rh)
		# + = turn left, - = turn right (Ru)
		# input angles are in degrees, functions need radians
		
		sring <- unlist(strsplit(string[which[n]], ""))	

		for (i in 1:nrow(drules)) { # translate to drawing commands
			for (j in 1:length(sring)) {
					if (sring[j] == drules$symbol[i]) sring[j] <- drules$action[i]
					}		
			}

		fifo <- vector("list") # store info to restore later
		ns <- 0L # stack counter
			
		# execute the drawing instructions
		cp <- c(0, stepSize, 0) # cp = current position
		# assumes there was a hidden step from 0,0,0 along H axis
		# the H axis is the +y axis in rgl
		# necessary b/c rotation matrices will give 0,0,0 otherwise
		
		open3d(...)

		for (j in 1:length(sring))	{
			#cat("Processing character", j, "\n")
			if (sring[j] == "F") {
				 x <- cp[1] + stepSize*cos(ch*pi/180)
				 y <- cp[2] + stepSize*sin(ch*pi/180)
				 grid.line.to(x, y, default.units = "native", ...)
				 cp <- c(x, y)
				 }

			if (sring[j] == "f") {
				 x <- cp[1] + stepSize*cos(ch*pi/180)
				 y <- cp[2] + stepSize*sin(ch*pi/180)
				 grid.move.to(x, y, default.units = "native")
				 cp <- c(x, y)
				 }

			if (sring[j] == "[") {
				#cat("Found a [ \n")
				#cat("ns is:", ns, "\n")
				ns <- ns + 1 # save the current settings
				fifo[[ns]] <- c(cp, ch)
				#print(fifo)			
				}

			if (sring[j] == "]") {
				#cat("Found a ] \n")
				#cat("ns is:", ns, "\n")
				cp <- fifo[[ns]][1:2]
				ch <- fifo[[ns]][3]
				grid.move.to(cp[1], cp[2], default.units = "native")
				ns <- ns - 1
				#print(fifo)			
				}

			if (sring[j] == "+") cp <- cp %*% Ru(angle*pi/180)

			if (sring[j] == "-") cp <- cp %*% Ru(-1*angle*pi/180)

			if (sring[j] == "&") cp <- cp %*% Rl(angle*pi/180)

			if (sring[j] == "^") cp <- cp %*% Rl(-1*angle*pi/180)

			if (sring[j] == "\\") cp <- cp %*% Rh(angle*pi/180)

			if (sring[j] == "/") cp <- cp %*% Rh(-1*angle*pi/180)

			if (sring[j] == "|") cp <- cp %*% Ru(pi) # turn around

			
			}
		} # end of looping over which
		
	}
	

R/rotationMatrices.R
##' 3D Rotation Matrices for use with Lindenmayer Systems
##'
##' These functions generate rotation matrices for use with Lindenmayer or L-Systems.
##' Called by \code{\link{drawLsys3D}} and not normally by the user.
##'
##' @param a Numeric.  An angle in radians.
##'
##' @return A 3 x 3 Matrix.
##'
##' @section Details These matrices correspond to a right-hand system, with (+)-ive
##' rotation being clockwise from the reference axis.  See
##' the coordinate system diagram in the README.  These matrices should
##' premultiply the coordinates to get the new coordinates.
##'
##' @name Rh-Rl-Ru
##' @rdname Rh-Rl-Ru
##' @alias Ru
##' @alias Rl
##' @alias Rh
##' @export
##' @keywords utilities
##'

Rh <- function(a) { # rotation around x/H axis
	M <- c(1.0, 0.0, 0.0, 0.0, cos(a), sin(a), 0.0, -sin(a), cos(a))
	M <- matrix(M, ncol = 3, byrow = TRUE)
	M
	}
	
Rl <- function(a) { # rotation around y/L axis
	M <- c(cos(a), 0.0, sin(a), 0.0, 1.0, 0.0, -sin(a), 0.0, cos(a))
	M <- matrix(M, ncol = 3, byrow = TRUE)
	M
	}
	
Ru <- function(a) { # rotation around z/U axis
	M <- c(cos(a), -sin(a), 0.0, sin(a), cos(a), 0.0, 0.0, 0.0, 1.0)
	M <- matrix(M, ncol = 3, byrow = TRUE)
	M
	}



R/LindenmayeR-package.R
##' Functions to Explore L-Systems (Lindenmayer Systems)
##' 
##' Lindenmayer or L-systems are parallel rewriting systems
##' which can be used to simulate biological forms and certain
##' kinds of fractals.  Briefly, in an L-system a series of
##' symbols in a string are replaced iteratively according to
##' rules to give a more complex string. Eventually, the
##' symbols are translated into turtle
##' graphics for plotting.  Wikipedia has a very good
##' introduction: \url{https://en.wikipedia.org/wiki/L-system}
##' This package provides basic functions for exploring L-systems.
##'
##' @name LindenmayeR-package
##' @title Lindenmayer System Functions
##' @docType package
##' @author Bryan A. Hanson
##' @import stringr
##' @import grid
##' @rdname LindenmayeR-package
##' @keywords package
NULL


R/drawLsys.R
##' Draw a 2D L-System Using Turtle Graphics
##'
##' This function takes input strings, previously created with \code{\link{Lsys}},
##' translates them into 2D turtle graphics instructions, and then plots the results.
##'
##' @param string A character vector giving the strings containing the turtle graphics
##' instructions.  Created by \code{\link{Lsys}}.  The "language" and character set
##' of this string is arbitary.  Compare the examples below for the modified Koch
##' curve and the Sierpinski triangle.
##'
##' @param drules A data frame containing columns "symbols" and "action".  These contain the input
##' symbols and the corresponding drawing action.  The symbol column is in the
##' character set used by \code{\link{Lsys}} and is arbitary.  The action column
##' entries must be from the set \code{c("F", "f", "+", "-", "[", "]")}.  These are
##' the final drawing instructions and are interpreted as follows:
##' \describe{
##'   \item{"F"}{Move forward drawing as you go.}
##'   \item{"f"}{Move forward w/o drawing.}
##'   \item{"+"}{Turn by positive \code{ang}.}
##'   \item{"-"}{Turn by negative \code{ang}.}
##'   \item{"["}{Save current position and heading.}
##'   \item{"]"}{Restore saved position and heading (allows one to go back).}
##' }
##' See the examples.  Note that the "action" entry always uses these symbols,
##' though not all of them need be used.
##'
##' @param st A numeric vector of length 3 giving the screen coordinates where
##' the start of the curve should be placed.  The screen is 100 x 100 with the
##' lower left corner as 0,0.  The third element is the initial drawing angle
##' in degrees.
##'
##' @param stepSize Numeric.  The length of the drawing step.
##'
##' @param ang Numeric.  The angle in degrees when a change in direction is requested.
##'
##' @param which Integer.  The entries in \code{string} which should be drawn.  Defaults
##' to the last (most complex) entry.  If \code{length(which) > 1} each plot is drawn in
##' its own window.
##'
##' @param shrinkFactor A numeric vector of the same length as \code{string}.  As each
##' plot is made, \code{stepSize} will be divided by the corresponding value in \code{shrinkFactor}.
##' This allows one to scale down the increasingly large/complex plots to make them
##' occupy a space similar to the less complex plots.
##'
##' @param ...  Additional parameters to be passed to the \code{grid} drawing routines.
##' Most likely, something of the form \code{gp = gpar(...)}.  See \code{\link{gpar}}
##' and the last example.
##'
##' @section Warning: Remember that if \code{retAll = TRUE}, \code{\link{Lsys}} returns
##' the initial string plus the results of all iterations.  In this case, if you want
##' the 5th iteration, you should specify \code{which = 6} since
##' the initial string is in \code{string[1]}.
##'
##' @return none, side effect is a plot, UNLESS record is set to be TURE then a data frame, containing coordinate data and segment types
##'   s for start of a segment d for segment endpoints to be drawn, r for reverse points, see \code{\link{recLsys}}
##'
##' @name drawLsys
##' @rdname drawLsys
##' @export
##' @keywords plot
##'
##' @examples
##' require('grid')
##'
##' # Modified Koch curve
##' rkoch1 <- data.frame(inp = c("F"), out = c("F+F-F-F+F"), stringsAsFactors = FALSE)
##' k1 <- Lsys(init = "F", rules = rkoch1, n = 3)
##' dkoch <- data.frame(symbol = c("F", "f", "+", "-", "[", "]"),
##' action = c("F", "f", "+", "-", "[", "]"), stringsAsFactors = FALSE)
##' drawLsys(string = k1, stepSize = 3, st = c(10, 50, 0), drules = dkoch)
##' grid.text("Modified Koch Curve (n = 3)", 0.5, 0.25)
##'
##' # Classic Koch snowflake
##' rkoch2 <- data.frame(inp = c("F"), out = c("F-F++F-F"), stringsAsFactors = FALSE)
##' k2 <- Lsys(init = "F++F++F", rules = rkoch2, n = 4)
##' drawLsys(string = k2, stepSize = 1, ang = 60, st = c(10, 25, 0), drules = dkoch)
##' grid.text("Classic Koch Snowflake (n = 4)", 0.5, 0.5)
##'
##' # Sierpinski Triangle
##' rSierp <- data.frame(inp = c("A", "B"), out = c("B-A-B", "A+B+A"), stringsAsFactors = FALSE)
##' s <- Lsys(init = "A", rules = rSierp, n = 6)
##' dSierp <- data.frame(symbol = c("A", "B", "+", "-", "[", "]"),
##' action = c("F", "F", "+", "-", "[", "]"), stringsAsFactors = FALSE)
##' drawLsys(string = s, stepSize = 1, ang = 60, st = c(20, 25, 0), drules = dSierp)
##' grid.text("Sierpinski Triangle (n = 6)", 0.5, 0.1)
##'
##' # Islands & Lakes
##' islands_rules <- data.frame(inp = c("F", "f"), out = c("F+f-FF+F+FF+Ff+FF-f+FF-F-FF-Ff-FFF",
##' "ffffff"), stringsAsFactors = FALSE)
##' islands <- Lsys(init = "F+F+F+F", rules = islands_rules, n = 2)
##' draw_islands <- data.frame(symbol = c("F", "f", "+", "-", "[", "]"),
##' action = c("F", "f", "+", "-", "[", "]"), stringsAsFactors = FALSE)
##' drawLsys(string = islands, step = 1, ang = 90, st = c(70, 35, 90),
##' drules = draw_islands,  gp = gpar(col = "red", fill = "gray"))
##'
##' # A primitive tree (aka Pythagoras Tree)
##' prim_rules <- data.frame(inp = c("0", "1"),
##' out = c("1[+0]-0", "11"), stringsAsFactors = FALSE)
##' primitive_plant <- Lsys(init = "0", rules = prim_rules, n = 7)
##' draw_prim <- data.frame(symbol = c("0", "1", "+", "-", "[", "]"),
##' action = c("F", "F", "+", "-", "[", "]"), stringsAsFactors = FALSE)
##' drawLsys(string = primitive_plant, stepSize = 1, ang = 45, st = c(50, 5, 90),
##' drules = draw_prim, which = 7)
##' grid.text("Primitive Tree (n = 6)", 0.5, 0.75)
##'
##' # A more realistic botanical structure
##' # Some call this a fractal tree, I think it is more like seaweed
##' # Try drawing the last iteration (too slow for here, but looks great)
##' fractal_tree_rules <- data.frame(inp = c("X", "F"),
##' out = c("F-[[X]+X]+F[+FX]-X", "FF"), stringsAsFactors = FALSE)
##' fractal_tree <- Lsys(init = "X", rules = fractal_tree_rules, n = 7)
##' draw_ft <- data.frame(symbol = c("X", "F", "+", "-", "[", "]"),
##' action = c("f", "F", "+", "-", "[", "]"), stringsAsFactors = FALSE)
##' drawLsys(string = fractal_tree, stepSize = 2, ang = 25, st = c(50, 5, 90),
##' drules = draw_ft, which = 5, gp = gpar(col = "chocolate4", fill = "honeydew"))
##' grid.text("Fractal Seaweed (n = 4)", 0.25, 0.25)
##'
##' # Calculation/rendering separation
##'
##' d = drawLsys(fractal_tree, stepSize = 2, ang = 25, st = c(50, 5, 90),drules = draw_ft, which = 5,record=TRUE)
##' ggLsys(d)
##'
##'

drawLsys <- function(string = NULL, drules = NULL,
	st = c(5, 50, 0), stepSize = 1.0, ang = 90.0,
	which = length(string), shrinkFactor = NULL,
	record = FALSE,
	...) {

	# check drules to make sure only allowed characters were given
	OK <- c("F", "f", "+", "-", "[", "]")
	test <- drules$action
	if (!all(test %in% OK)) {
		msg1 <- paste("Only the following actions are recognized:",
			paste(OK, collapse = " "), sep = " ")
		message(msg1)
		good <- test %in% OK
		bad <- test[!good]
		msg2 <- paste("I can't use these:",
			paste(bad, collapse = " "), sep = " ")
		stop(msg2)
		}

	for (n in 1:length(which)) {

		# set up the viewport
		grid.newpage()
		vp <- viewport(x = 0.5, y = 0.5,
			width = 1.0, height = 1.0,
			xscale = c(0, 100), yscale = c(0, 100),
			name = "VP0")
		pushViewport(vp)
		grid.rect(...) # needed if a colored bkgnd is desired

		# convert the initial string into drawing instructions
		# F = move forward; f = move w/o drawing
		# +, - turn by angle; [ save cp, ch; ] restore saved cp, ch

		string <- unlist(strsplit(string[which[n]], ""))

		for (i in 1:nrow(drules)) { # translate to drawing commands
			for (j in 1:length(string)) {
					if (string[j] == drules$symbol[i]) string[j] <- drules$action[i]
					}
			}

		# execute the drawing instructions

		grid.move.to(st[1], st[2], default.units = "native")
		cp <- st # cp = current point
		ch <- st[3] # ch = current heading, 0 = East in degrees
		if (!is.null(shrinkFactor)) stepSize <- stepSize/shrinkFactor
		fifo <- vector("list") # store info to restore later
		ns <- 0L # stack counter



		if(isTRUE(record)){
		  recLsys(string = string, drules =drules,
		                     st =st, stepSize =stepSize, ang =ang,
		                     which = which, shrinkFactor =shrinkFactor)
		}else{
		  for (j in 1:length(string))	{
		    #cat("Processing character", j, "\n")
		    if (string[j] == "F") {
		      x <- cp[1] + stepSize*cos(ch*pi/180)
		      y <- cp[2] + stepSize*sin(ch*pi/180)
		      grid.line.to(x, y, default.units = "native", ...)
		      cp <- c(x, y)
		    }else	if(string[j] == "f") {
		      x <- cp[1] + stepSize*cos(ch*pi/180)
		      y <- cp[2] + stepSize*sin(ch*pi/180)
		      grid.move.to(x, y, default.units = "native")
		      cp <- c(x, y)
		    }else if(string[j] == "[") {
		      #cat("Found a [ \n")
		      #cat("ns is:", ns, "\n")
		      ns <- ns + 1 # save the current settings
		      fifo[[ns]] <- c(cp, ch)
		      #print(fifo)
		    }else if(string[j] == "]") {
		      #cat("Found a ] \n")
		      #cat("ns is:", ns, "\n")
		      cp <- fifo[[ns]][1:2]
		      ch <- fifo[[ns]][3]
		      grid.move.to(cp[1], cp[2], default.units = "native")
		      ns <- ns - 1
		      #print(fifo)
		    }else	if(string[j] == "-"){
		      ch = ch - ang
		    }else{
		      if (string[j] == "+") ch = ch + ang
		    }
		  }
		} # end of looping over which
		}


	}
